File: SamplePresence/UIKit-ObjC/JoinViewController.m
#import "JoinViewController.h"
#import "DemoFlowController.h"
#import "FormFieldCell.h"

typedef NS_ENUM(NSInteger, JoinSection) {
    JoinSectionSession = 0,
    JoinSectionLocalUser = 1,
    JoinSectionAPIKey = 2,
    JoinSectionStatus = 3,
    JoinSectionJoin = 4,
    JoinSectionCount = 5
};

@interface JoinViewController ()
@property (nonatomic, weak) DemoFlowController *flowController;
@end

@implementation JoinViewController

- (instancetype)initWithFlowController:(DemoFlowController *)flowController {
    self = [super initWithStyle:UITableViewStyleInsetGrouped];
    if (self) {
        _flowController = flowController;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Join Call";
    [self.tableView registerClass:[FormFieldCell class] forCellReuseIdentifier:FormFieldCell.reuseIdentifier];
    [self.tableView registerClass:[FormMultilineCell class] forCellReuseIdentifier:FormMultilineCell.reuseIdentifier];
    [self.tableView registerClass:[FormSwitchCell class] forCellReuseIdentifier:FormSwitchCell.reuseIdentifier];
    [self bindFlowController];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.flowController.errorMessage == nil ? JoinSectionCount : JoinSectionCount + 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    if (section >= JoinSectionCount) { return nil; }
    switch (section) {
        case JoinSectionSession: return @"Session";
        case JoinSectionLocalUser: return @"Local User";
        case JoinSectionAPIKey: return @"API Key";
        case JoinSectionStatus: return @"Call Status";
        default: return nil;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (section == JoinSectionCount) { return 1; }
    if (section == JoinSectionStatus) {
        return self.flowController.callStatusMessage.length > 0 ? 1 : 0;
    }
    switch (section) {
        case JoinSectionSession: return 4;
        case JoinSectionLocalUser: return 4;
        case JoinSectionAPIKey: return 1;
        case JoinSectionJoin: return 1;
        default: return 0;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == JoinSectionCount) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ErrorCell"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"ErrorCell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            cell.textLabel.numberOfLines = 0;
        }
        cell.textLabel.text = self.flowController.errorMessage;
        cell.textLabel.textColor = UIColor.systemRedColor;
        return cell;
    }

    DemoSession *session = self.flowController.session;

    if (indexPath.section == JoinSectionSession) {
        if (indexPath.row == 3) {
            FormMultilineCell *cell = [tableView dequeueReusableCellWithIdentifier:FormMultilineCell.reuseIdentifier forIndexPath:indexPath];
            [cell configureWithTitle:@"Session Token:" text:session.sessionToken];
            __weak DemoSession *weakSession = session;
            cell.onTextChanged = ^(NSString *text) {
                weakSession.sessionToken = text ?: @"";
            };
            return cell;
        }
        FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
        switch (indexPath.row) {
            case 0:
                [cell configureWithTitle:@"Session ID:" text:session.sessionID placeholder:@"Required" keyboardType:UIKeyboardTypeDefault];
                [cell.valueField addTarget:self action:@selector(sessionIDChanged:) forControlEvents:UIControlEventEditingChanged];
                break;
            case 1:
                [cell configureWithTitle:@"PIN:" text:session.sessionPIN placeholder:@"Required" keyboardType:UIKeyboardTypeDefault];
                [cell.valueField addTarget:self action:@selector(sessionPINChanged:) forControlEvents:UIControlEventEditingChanged];
                break;
            case 2:
                [cell configureWithTitle:@"GSS URL:" text:session.gssServerURL placeholder:@"gss+ssl://…" keyboardType:UIKeyboardTypeURL];
                [cell.valueField addTarget:self action:@selector(gssURLChanged:) forControlEvents:UIControlEventEditingChanged];
                break;
            default:
                break;
        }
        return cell;
    }

    if (indexPath.section == JoinSectionLocalUser) {
        if (indexPath.row == 2 || indexPath.row == 3) {
            FormSwitchCell *cell = [tableView dequeueReusableCellWithIdentifier:FormSwitchCell.reuseIdentifier forIndexPath:indexPath];
            if (indexPath.row == 2) {
                [cell configureWithTitle:@"Camera On" isOn:session.cameraEnabled];
                [cell.toggle addTarget:self action:@selector(cameraChanged:) forControlEvents:UIControlEventValueChanged];
            } else {
                [cell configureWithTitle:@"Microphone On" isOn:session.microphoneEnabled];
                [cell.toggle addTarget:self action:@selector(microphoneChanged:) forControlEvents:UIControlEventValueChanged];
            }
            return cell;
        }
        FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
        if (indexPath.row == 0) {
            [cell configureWithTitle:@"Display Name:" text:session.displayName placeholder:@"Required" keyboardType:UIKeyboardTypeDefault];
            [cell.valueField addTarget:self action:@selector(displayNameChanged:) forControlEvents:UIControlEventEditingChanged];
        } else {
            [cell configureWithTitle:@"Avatar URL:" text:session.avatarURL placeholder:@"https://…" keyboardType:UIKeyboardTypeURL];
            [cell.valueField addTarget:self action:@selector(avatarURLChanged:) forControlEvents:UIControlEventEditingChanged];
        }
        return cell;
    }

    if (indexPath.section == JoinSectionAPIKey) {
        FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
        [cell configureWithTitle:@"Help Lightning API Key:" text:session.apiKey placeholder:@"Required" keyboardType:UIKeyboardTypeDefault];
        [cell.valueField addTarget:self action:@selector(apiKeyChanged:) forControlEvents:UIControlEventEditingChanged];
        return cell;
    }

    if (indexPath.section == JoinSectionStatus) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"StatusCell"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"StatusCell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            cell.textLabel.numberOfLines = 0;
        }
        cell.textLabel.text = self.flowController.callStatusMessage;
        cell.textLabel.textColor = self.flowController.callPhase == DemoCallPhaseEnded
            ? UIColor.systemRedColor
            : UIColor.labelColor;
        return cell;
    }

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"JoinCell"];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"JoinCell"];
    }
    UIListContentConfiguration *configuration = [UIListContentConfiguration cellConfiguration];
    configuration.text = self.flowController.isBusy ? @"Joining…" : @"Join Call";
    configuration.textProperties.alignment = NSTextAlignmentCenter;
    configuration.textProperties.color = self.flowController.isBusy ? UIColor.secondaryLabelColor : self.view.tintColor;
    cell.contentConfiguration = configuration;
    cell.selectionStyle = self.flowController.isBusy ? UITableViewCellSelectionStyleNone : UITableViewCellSelectionStyleDefault;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (indexPath.section == JoinSectionJoin && !self.flowController.isBusy) {
        [self.flowController joinCallFromViewController:self];
    }
}

- (void)sessionIDChanged:(UITextField *)field { self.flowController.session.sessionID = field.text ?: @""; }
- (void)sessionPINChanged:(UITextField *)field { self.flowController.session.sessionPIN = field.text ?: @""; }
- (void)gssURLChanged:(UITextField *)field { self.flowController.session.gssServerURL = field.text ?: @""; }
- (void)displayNameChanged:(UITextField *)field { self.flowController.session.displayName = field.text ?: @""; }
- (void)avatarURLChanged:(UITextField *)field { self.flowController.session.avatarURL = field.text ?: @""; }
- (void)apiKeyChanged:(UITextField *)field { self.flowController.session.apiKey = field.text ?: @""; }
- (void)cameraChanged:(UISwitch *)toggle { self.flowController.session.cameraEnabled = toggle.isOn; }
- (void)microphoneChanged:(UISwitch *)toggle { self.flowController.session.microphoneEnabled = toggle.isOn; }

- (void)bindFlowController {
    __weak typeof(self) weakSelf = self;
    self.flowController.onBusyChanged = ^(BOOL busy) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)busy;
        [strongSelf.tableView reloadData];
    };
    self.flowController.onErrorChanged = ^(NSString * _Nullable errorMessage) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)errorMessage;
        [strongSelf.tableView reloadData];
    };
    self.flowController.onCallStatusChanged = ^(NSString *message) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)message;
        [strongSelf.tableView reloadSections:[NSIndexSet indexSetWithIndex:JoinSectionStatus]
                            withRowAnimation:UITableViewRowAnimationAutomatic];
    };
    self.flowController.onCallPhaseChanged = ^(DemoCallPhase phase) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)phase;
        [strongSelf.tableView reloadData];
    };
}

@end

File: SamplePresence/UIKit/JoinViewController.swift
import UIKit
import HLSDK

final class JoinViewController: UITableViewController {
    private let flowController: DemoFlowController

    private enum Section: Int, CaseIterable {
        case session
        case localUser
        case apiKey
        case status
        case join
    }

    private enum SessionRow: Int, CaseIterable {
        case sessionID
        case pin
        case gssURL
        case sessionToken
    }

    private enum LocalUserRow: Int, CaseIterable {
        case displayName
        case avatarURL
        case camera
        case microphone
    }

    init(flowController: DemoFlowController) {
        self.flowController = flowController
        super.init(style: .insetGrouped)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Join Call"
        tableView.register(FormFieldCell.self, forCellReuseIdentifier: FormFieldCell.reuseIdentifier)
        tableView.register(FormMultilineCell.self, forCellReuseIdentifier: FormMultilineCell.reuseIdentifier)
        tableView.register(FormSwitchCell.self, forCellReuseIdentifier: FormSwitchCell.reuseIdentifier)
        bindFlowController()
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        var count = Section.allCases.count
        if flowController.errorMessage != nil {
            count += 1
        }
        return count
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch sectionIdentifier(for: section) {
        case .session: return "Session"
        case .localUser: return "Local User"
        case .apiKey: return "API Key"
        case .status: return "Call Status"
        case .join: return nil
        default: return nil
        }
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch sectionIdentifier(for: section) {
        case .session: return SessionRow.allCases.count
        case .localUser: return LocalUserRow.allCases.count
        case .apiKey: return 1
        case .status: return flowController.callStatusMessage.isEmpty ? 0 : 1
        case .join: return 1
        case .none: return 1
        }
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch sectionIdentifier(for: indexPath.section) {
        case .none:
            let cell = UITableViewCell(style: .default, reuseIdentifier: "ErrorCell")
            cell.selectionStyle = .none
            cell.textLabel?.text = flowController.errorMessage
            cell.textLabel?.textColor = .systemRed
            cell.textLabel?.numberOfLines = 0
            return cell
        case .session:
            return sessionCell(for: indexPath)
        case .localUser:
            return localUserCell(for: indexPath)
        case .apiKey:
            return apiKeyCell(for: indexPath)
        case .status:
            let cell = UITableViewCell(style: .default, reuseIdentifier: "StatusCell")
            cell.selectionStyle = .none
            cell.textLabel?.numberOfLines = 0
            if case .ended = flowController.callPhase {
                cell.textLabel?.textColor = .systemRed
            } else {
                cell.textLabel?.textColor = .label
            }
            cell.textLabel?.text = flowController.callStatusMessage
            return cell
        case .join:
            let cell = UITableViewCell(style: .default, reuseIdentifier: "JoinCell")
            var configuration = UIListContentConfiguration.cell()
            configuration.text = flowController.isBusy ? "Joining…" : "Join Call"
            configuration.textProperties.alignment = .center
            configuration.textProperties.color = flowController.isBusy ? .secondaryLabel : view.tintColor
            cell.contentConfiguration = configuration
            cell.selectionStyle = flowController.isBusy ? .none : .default
            return cell
        }
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        guard sectionIdentifier(for: indexPath.section) == .join, !flowController.isBusy else { return }
        Task { await flowController.joinCall(from: self) }
    }

    private func sectionIdentifier(for section: Int) -> Section? {
        if section < Section.allCases.count {
            return Section(rawValue: section)
        }
        return nil
    }

    private func sessionCell(for indexPath: IndexPath) -> UITableViewCell {
        guard let row = SessionRow(rawValue: indexPath.row) else { return UITableViewCell() }

        switch row {
        case .sessionToken:
            guard let cell = tableView.dequeueReusableCell(withIdentifier: FormMultilineCell.reuseIdentifier, for: indexPath) as? FormMultilineCell else {
                return UITableViewCell()
            }
            cell.configure(title: "Session Token:", text: flowController.session.sessionToken)
            cell.onTextChanged = { [weak self] text in
                self?.flowController.session.sessionToken = text
            }
            return cell
        default:
            guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell else {
                return UITableViewCell()
            }
            switch row {
            case .sessionID:
                cell.configure(title: "Session ID:", text: flowController.session.sessionID, placeholder: "Required")
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.sessionID = cell.valueField.text ?? ""
                }, for: .editingChanged)
            case .pin:
                cell.configure(title: "PIN:", text: flowController.session.sessionPIN, placeholder: "Required")
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.sessionPIN = cell.valueField.text ?? ""
                }, for: .editingChanged)
            case .gssURL:
                cell.configure(title: "GSS URL:", text: flowController.session.gssServerURL, placeholder: "gss+ssl://…")
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.gssServerURL = cell.valueField.text ?? ""
                }, for: .editingChanged)
            case .sessionToken:
                break
            }
            return cell
        }
    }

    private func localUserCell(for indexPath: IndexPath) -> UITableViewCell {
        guard let row = LocalUserRow(rawValue: indexPath.row) else { return UITableViewCell() }

        switch row {
        case .camera, .microphone:
            guard let cell = tableView.dequeueReusableCell(withIdentifier: FormSwitchCell.reuseIdentifier, for: indexPath) as? FormSwitchCell else {
                return UITableViewCell()
            }
            switch row {
            case .camera:
                cell.configure(title: "Camera On", isOn: flowController.session.cameraEnabled)
                cell.toggle.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.cameraEnabled = cell.toggle.isOn
                }, for: .valueChanged)
            case .microphone:
                cell.configure(title: "Microphone On", isOn: flowController.session.microphoneEnabled)
                cell.toggle.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.microphoneEnabled = cell.toggle.isOn
                }, for: .valueChanged)
            default:
                break
            }
            return cell
        default:
            guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell else {
                return UITableViewCell()
            }
            switch row {
            case .displayName:
                cell.configure(title: "Display Name:", text: flowController.session.displayName, placeholder: "Required")
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.displayName = cell.valueField.text ?? ""
                }, for: .editingChanged)
            case .avatarURL:
                cell.configure(title: "Avatar URL:", text: flowController.session.avatarURL, placeholder: "https://…")
                cell.valueField.addAction(UIAction { [weak self] _ in
                    self?.flowController.session.avatarURL = cell.valueField.text ?? ""
                }, for: .editingChanged)
            default:
                break
            }
            return cell
        }
    }

    private func apiKeyCell(for indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell else {
            return UITableViewCell()
        }
        cell.configure(title: "Help Lightning API Key:", text: flowController.session.apiKey, placeholder: "Required")
        cell.valueField.addAction(UIAction { [weak self] _ in
            self?.flowController.session.apiKey = cell.valueField.text ?? ""
        }, for: .editingChanged)
        return cell
    }

    private func bindFlowController() {
        flowController.onBusyChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
        flowController.onErrorChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
        flowController.onCallStatusChanged = { [weak self] _ in
            self?.tableView.reloadSections(IndexSet(integer: Section.status.rawValue), with: .automatic)
        }
        flowController.onCallPhaseChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
    }
}
